In [1]:
import bs4
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import requests
import math
from textblob import TextBlob
from wordcloud import WordCloud, STOPWORDS

    to fetch all the comments we will need to iterate through multiple pages of comments and to do this lets find the total number of pages

In [ ]:
import requests
import bs4
import math

url = "https://www.amazon.in/Apple-iPhone-XR-64GB-Black/product-reviews/B07JWV47JW/ref=cm_cr_dp_d_show_all_btm?ie=UTF8&reviewerType=all_reviews"

page = requests.get(url)
print(page.status_code)

soup = bs4.BeautifulSoup(page.content, 'html.parser')

# Check if the list has elements before accessing them
num = soup.find_all('span', {'data-hook': "cr-filter-info-review-count"})
if num:
    total = str(num[0].text)
    t = total[-9::-1].split()
    tt = t[0]
    tt = "".join(reversed(tt))
    tt = tt.replace(",", '')
    tt = math.ceil(int(tt) / 10)
    print(tt)
else:
    print("No review count found on the page.")


In [15]:
url = "https://www.amazon.in/Apple-iPhone-XR-64GB-Black/product-reviews/B07JWV47JW/ref=cm_cr_dp_d_show_all_btm?ie=UTF8&reviewerType=all_reviews"


page = requests.get(url)
print(page.status_code)
html = page.content

soup = bs4.BeautifulSoup(html, 'html.parser')
num = soup.find_all('span', {'data-hook':"cr-filter-info-review-count"})
total = str(num[0].text)
t = total[-9::-1].split()
tt = t[0]
tt = "".join(reversed(tt))
tt = tt.replace(",", '')
tt = math.ceil(int(tt)/10)
tt

503


IndexError: list index out of range

    Now fetching the name of the person who reviewed the product, star rating given by that person, comment title and the comment and storing them in lists so that they can further be used for our analysis

In [ ]:
r_name = []
r_title = []
r_body = []
r_star = []
for pn in range(1, tt+1):
    #global soup, r_name, link
    #if soup.find_all('li', class_="a-disabled a-last"):
    #print(f"page no {pn}")
    #else:
        
    #print("andar ghus gaye")
    #pn += 1
    #print(pn)
    if pn > 1:
        link = f"https://www.amazon.in/Apple-iPhone-XR-64GB-Black/product-reviews/B07JWV47JW/ref=cm_cr_arp_d_paging_btm_next_{pn}?ie=UTF8&reviewerType=all_reviews&pageNumber={pn}"

        page = requests.get(link)

    #page.status_code

        #page.reason

        #page.headers['Content-type']
    if page.status_code == 200:
        print(f"page no {pn}")
        html = page.content

        soup = bs4.BeautifulSoup(html, 'html.parser')


        comm = soup.find_all('span', attrs = {'class' : cls})
        ns = 10*(pn-1)
        ne = ns+len(comm)
        for i in range(0, len(comm)):
            
            r_name.append(comm[i].text)
            #next_page(pn)

        title = soup.find_all('a', class_="review-title-content")
        for i in range(0, len(title)):
            r_title.append(title[i].text)
        r_title[:] = [text.strip('\n') for text in r_title]
        star = soup.find_all('i', class_="review-rating")
        for i in range(0, len(star)):
            r_star.append(star[i].text)
        body = soup.find_all('span', class_="review-text-content")
        for i in range(0, len(body)):
            r_body.append(body[i].text)
        r_body[:] = [text.strip('\n') for text in r_body]

#r_name

In [ ]:
r_name

    as we can see from the above example, we have some duplicacies in data which are caused due to the speific structure of that web page.
    
    Let's remove these duplication and make to make our analysis better and more refined.
    
    Since there can be fare chance that 2 person might have same first name, hence we will first put this data in a pandas dataframe and then remove duplicates in form of pairs assuming that 2 persons with same first and lastname might not have given exact same rating. 

In [3]:
r_starrate = []
for i in r_star:
    r = str(i)
    r_starrate.append(r[2]) 

NameError: name 'r_star' is not defined

In [4]:
df = pd.DataFrame()

In [5]:
df['Author'] = r_name
df['Rating'] = r_starrate

NameError: name 'r_name' is not defined

In [6]:
df

Empty DataFrame
Columns: []
Index: []

In [7]:
df.drop_duplicates(inplace = True)

Empty DataFrame
Columns: []
Index: []

In [8]:
df.to_csv('ratings.csv', index = False)

In [9]:
#df.groupby("Rating").count().plot(kind = 'bar')
dd = df.groupby("Rating").count()
dd.plot(kind = 'area', subplots = True, figsize = (15, 8), title = "Analysis of Ratings given for the given product in terms of", legend = False, xticks = [1, 2,3, 4, 5], rot = (0, 0), colormap = 'plasma')
plt.xlabel("Rating")
plt.xticks(rotation = 90)
plt.ylabel("Number of AUthors")
plt.show()

KeyError: 'Rating'

    
    
   This area plot clearly depicts that `maximum` users have given a `5 star` rating for this product and tells a lot about overall impression of this product on users

In [ ]:
dff = pd.DataFrame()
dff['Comment_Title'] = r_title
dff['Review'] = r_body

In [ ]:
dff

In [ ]:
dff.drop_duplicates(inplace = True)

In [ ]:
dff.to_csv("reviews.csv", index = False)

In [ ]:
from wordcloud import WordCloud, STOPWORDS

In [ ]:
com = pd.read_csv("reviews.csv")


In [ ]:
co = com['Comment_Title']

In [ ]:
co

In [ ]:
st = co.to_string(index = False, header = None)

In [ ]:
stop = set(STOPWORDS)
wc = WordCloud(width = 2560, height = 1440, stopwords = stop, min_font_size = 10, max_words = 100, background_color='white').generate(st)
wc.to_file('wc.png')
plt.title("Analysis of comments title based on word frequency")
plt.imshow(wc)
plt.show()

    This shows that people usually believe that this product is good, excellent, Awesome, best, fantastic, nice product, etc and tells us that people are usually have with this product.

In [ ]:
c = com['Review']

In [10]:
c

NameError: name 'c' is not defined

In [11]:
s = c.to_string(index = False, header = None)

NameError: name 'c' is not defined

In [12]:
stop = set(STOPWORDS)
wc = WordCloud(width = 2560, height = 1440, stopwords = stop, min_font_size = 10, max_words = 200, background_color='white', colormap='cividis').generate(s)
wc.to_file('comment.png')
plt.title("Analysis of comments based on word frequency")
plt.imshow(wc)
plt.show()

NameError: name 's' is not defined

    Here we can see that this is much similar to that of the review title but few more things are added here. Here people also prefer to talk about some of its features like battery life, camera quality, fingerprint, display, face id, charging , etc.

## Sentimental Analysis

In [13]:
sent = TextBlob(s)
print(sent.sentiment)

NameError: name 's' is not defined

    Where subjectivity --> personal feelings, views, or beliefs.
                                (+0.0 => +1.0)
                                If subjectivity is near 1 it means that comment is more of a persoanal belief than a factual information
                                
              polarity --> emotions expressed in a sentence.
                              It tells if a comment is positive, negetive or neutral
                              (-1.0 => +1.0)

here it tells that overall comments live from `neutral to positive` for this product and is more based on `personal beliefs`

In [14]:
sent = TextBlob(st)
print(sent.sentiment)

NameError: name 'st' is not defined

now for sentiments in title part , titles are comparatively `more positive` for the comments and have `personal feelings`

### Dependencies

1. Execution time of this code highly depends on the number of reviews on that product and is directly propotional to it.

2. Mutiple Requests/Hits on Amazon may result in 503 or Server Unavailable error many times may be due to some security protocols followed by amazon

3. Disabling keepalive or deleting response object can be a good practise in this scenerio.

### Future Scope

1.    As this is only the backend focus project hence more work can be done on making a UI for this.

2. This project can be made more user friendly and dynamic by allowing user to enter product name from their end.